lib/pavilion/lockfile.py

```diff
@@ -0,0 +1,191 @@
+                        expires = time.time() + NEVER
+
+                if expires < time.time():
+                    # The file is expired. Try to delete it.
+                    try:
+
+                        with LockFile(self._lock_path + '.expired', expires_after=NEVER):
+                            try:
+                                os.unlink(self._lock_path)
+                            except OSError:
+                                pass
+                    except TimeoutError:
+                        pass
+
+                    # Only try this once.
+                    expires = time.time() + NEVER
+                else:
+                    time.sleep(self.SLEEP_PERIOD)
+
+        if not acquired:
+            raise TimeoutError("Lock on file '{}' could not be acquired.".format(self._lock_path))
+
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """Delete the lockfile, thereby releasing the lock.
+        :raises RuntimeError: When we can't delete our own lockfile for some reason.
+        """
+
+        # There isn't really anything we can do in this case.
+        host, user, expiration, lock_id = self.read_lockfile()
+
+        if lock_id is not None and lock_id != self._id:
+            LOGGER.error("Lockfile '{}' mysteriously replaced with one from {}."
+                         .format(self._lock_path, (host, user)))
+        else:
+            try:
+                os.unlink(self._lock_path)
+            except OSError as err:
+                # There isn't really anything we can do in this case.
+                host, user, expiration, lock_id = self.read_lockfile()
+
+                if lock_id == self._id:
+                    LOGGER.warning("Lockfile '{}' could not be deleted: '{}'"
+                                   .format(self._lock_path, err))
+                else:
+                    LOGGER.error("Lockfile '{}' mysteriously disappeared."
+                                 .format(self._lock_path))
+
+    @classmethod
+    def _create_lockfile(cls, path, expires, lock_id, group_id=None):
+        """Create and fill out a lockfile at the given path.
+        :param unicode path: Where the file will be created.
+        :param int expires: How far in the future the lockfile expires.
+        :param unicode lock_id: The unique identifier for this lockfile.
+        :returns: None
+        :raises IOError: When the file cannot be written too.
+        :raises OSError: When the file cannot be opened or already exists."""
+
+        # DEV NOTE: This logic is separate so that we can create these files outside of
+        # the standard mechanisms for testing purposes.
+
+        fd = os.open(path, os.O_EXCL | os.O_CREAT | os.O_RDWR)
+        expiration = time.time() + expires
+        file_note = b"{},{},{},{}".format(os.uname()[1], os.getlogin(), expiration, lock_id)
+        os.write(fd, file_note)
+        os.close(fd)
+
+        try:
+            os.chmod(path, cls.LOCK_PERMS)
```

Comment:
Is this tested?  Getting the right permissions took me a while.

Reply:
It's tested, and something I've done a bunch of times.  I use this type of synchronization pretty frequently.